File: python/SolverAiClientCompute.py
import requests
import json

from SolverAiComputeInput import SolverAiComputeInput
from SolverAiComputeResults import SolverAiComputeResults


class SolverAiClientCompute:

    def __init__(self, computerUrl: str, token: str, problemId: int) -> None:
        self.__base_url_Computer = computerUrl + "/"
        self.__problemId = problemId
        self.__headers = {
            "Authorization": f"Token {token}"
        }

    @staticmethod
    def __isStatusCodeOk(response):
        statusCode = response.status_code
        return 200 <= statusCode and statusCode < 300

    def getProblemSetup(self):
        headers = self.__headers.copy()
        headers["Content-Type"] = "application/json"
        url = f'{self.__base_url_Computer}problem_setup/{self.__problemId}'
        response = requests.get(url, headers=headers)
        if self.__isStatusCodeOk(response):
            try:
                data = json.loads(response.text)
                return data['inputs'], data['outputs']
            except Exception:
                raise Exception('Failed retrieving data.')
        else:
            raise Exception(f'Failed with code: {json.loads(response.text)}.')

    def runSolver(self, input: SolverAiComputeInput) -> SolverAiComputeResults:
        headers = self.__headers.copy()
        headers["Content-Type"] = "application/json"
        url = f'{self.__base_url_Computer}solvejson/'
        jsonData = input.getJson()
        response = requests.post(url, headers=headers, data=jsonData)
        if self.__isStatusCodeOk(response):
            try:
                data = json.loads(response.text)
            except Exception:
                return Exception('Failed retrieving data.')
            return SolverAiComputeResults(data['results'])
        else:
            return Exception(f'Failed with code: {json.loads(response.text)}.')

File: js/SolverAiClientCompute.js
const axios = require('axios');

const SolverAiComputeResults = require('./SolverAiComputeResults');


class SolverAiClientCompute {

    constructor(computerUrl, token, problemId) {
        this.__base_url_Computer = computerUrl + "/";
        this.__problemId = problemId;
        this.__headers = {
            Authorization: `Token ${token}`,
            "Content-Type": "application/json"
        };
    }

    static __isStatusCodeOk(statusCode) {
        return 200 <= statusCode && statusCode < 300;
    }

    async getProblemSetup() {
        try {
            const url = `${this.__base_url_Computer}problem_setup/${this.__problemId}`;
            const response = await axios.get(url, { headers: this.__headers });
            
            if (this.constructor.__isStatusCodeOk(response.status)) {
                return [response.data['inputs'], response.data['outputs']];
            } else {
                throw new Error(`Failed with code: ${response.data}`);
            }
        } catch (error) {
            throw new Error(`Exception: ${error.message} - ${error.response.statusText}`);
        }
    }

    async runSolver(input) {
        let inputJson = input.getJson();
        try {
            const url = `${this.__base_url_Computer}solvejson/`;
            const response = await axios.post(url, inputJson, { headers: this.__headers });
            
            if (this.constructor.__isStatusCodeOk(response.status)) {
                return new SolverAiComputeResults(response.data.results)
            } else {
                throw new Error(`Failed with code: ${response.data}`);
            }
        } catch (error) {
            throw new Error(`Exception: ${error.message}`);
        }
    }
}

module.exports = SolverAiClientCompute;
